pairtools/pairtools_dedup.py

```diff
@@ -1,248 +1,421 @@
+    elif (
+        pathlib.Path(output_unmapped).absolute() == pathlib.Path(output_dups).absolute()
+    ):
         outstream_unmapped = outstream_dups
     else:
-        outstream_unmapped = _fileio.auto_open(output_unmapped, mode='w', 
-                                            nproc=kwargs.get('nproc_out'),
-                                            command=kwargs.get('cmd_out', None))
-                             
+        outstream_unmapped = _fileio.auto_open(
+            output_unmapped,
+            mode="w",
+            nproc=kwargs.get("nproc_out"),
+            command=kwargs.get("cmd_out", None),
+        )
 
     header, body_stream = _headerops.get_header(instream)
     header = _headerops.append_new_pg(header, ID=UTIL_NAME, PN=UTIL_NAME)
-
     if send_header_to_dedup:
-        outstream.writelines((l+'\n' for l in header))
+        outstream.writelines((l + "\n" for l in header))
     if send_header_to_dup and outstream_dups and (outstream_dups != outstream):
-        outstream_dups.writelines((l+'\n' for l in header))
-    if (outstream_unmapped and (outstream_unmapped != outstream) 
-            and (outstream_unmapped != outstream_dups)):
-        outstream_unmapped.writelines((l+'\n' for l in header))
+        dups_header = header
+        dups_header[-1] += " parent_readID"
+        outstream_dups.writelines((l + "\n" for l in dups_header))
+    if (
+        outstream_unmapped
+        and (outstream_unmapped != outstream)
+        and (outstream_unmapped != outstream_dups)
+    ):
+        outstream_unmapped.writelines((l + "\n" for l in header))
 
     column_names = _headerops.extract_column_names(header)
     extra_cols1 = []
     extra_cols2 = []
     if extra_col_pair is not None:
         for col1, col2 in extra_col_pair:
             extra_cols1.append(
-                int(col1) if col1.isdigit() else column_names.index(col1))
+                int(col1) if col1.isdigit() else column_names.index(col1)
+            )
             extra_cols2.append(
-                int(col2) if col2.isdigit() else column_names.index(col2))
-                            
-
-    streaming_dedup( method, max_mismatch, sep,
-        c1, c2, p1, p2, s1, s2, extra_cols1, extra_cols2, unmapped_chrom,
-        body_stream, outstream, outstream_dups,
-        outstream_unmapped, out_stat, mark_dups)
+                int(col2) if col2.isdigit() else column_names.index(col2)
+            )
+
+    if backend == "cython":
+        streaming_dedup(
+            method,
+            max_mismatch,
+            sep,
+            c1,
+            c2,
+            p1,
+            p2,
+            s1,
+            s2,
+            extra_cols1,
+            extra_cols2,
+            unmapped_chrom,
+            body_stream,
+            outstream,
+            outstream_dups,
+            outstream_unmapped,
+            out_stat,
+            mark_dups,
+        )
+    elif backend in ("scipy", "sklearn"):
+        streaming_dedup_by_chunk(
```

Comment:
Ilya, does it make sense to create a function "streaming_dedup" which will do the same job as streaing_dedup_by_chunk but with no chunking? That will help us assess the typical number of errors per dataset due to chunking. 
If not, then probably we can safely rename this function to streaming_dedup.

Reply:
> Does anyone have thoughts regarding relying on the file header containing information about column names? It would make life much easier, but that's significantly breaking away from providing the order of columns as CLI arguments... Are there any pair files out there that don't have a header?

Yes, it looks like HiC-Pro[ does not rely on the header](http://nservant.github.io/HiC-Pro/RESULTS.html#list-of-valid-interaction-products). However, I'm not sure that existence of output of other tools with no header is a big trouble here. We typically generate the .pairs file with `pairtools parse`, and we always keep the header, as far as I know. For the files that are non-pairtools, we might provide a function `pairtools add_header` that will add user-specified order of columns to the header. Since pairtools have a [whole set of functions for manipulating the headers](http://nservant.github.io/HiC-Pro/RESULTS.html#list-of-valid-interaction-products), and it might be easy to create such a function.  
Any other opinions? @golobor

Reply:
I'd be in favour of generally relying on headers. Maybe we should decide and document the default order of columns in case the header is missing, and also I like the idea about `add_header`!